### PR TITLE
Add OpenAPI extensions to openapi_gen (Definition Generator)

### LINKED ISSUE #40818: Support for defining swagger extensions directly in types.go
Background:
kubectl and the Kubernetes reference docs both depend on the swagger.json spec to operate on Kubernetes resources.  The swagger.json lacks important information required by kubectl and the
docs.

Usecase:

- kubectl can identify the patchStrategy and mergeKey for resource fields
- kubectl can identify the columns to print for `kubectl get <type>` on resources
- docs can parse and display the default value for fields
- docs can identify top-level objects

Solution:

Add items to the comment tag: `// +k8s:openapi-gen`

Example:

```
// +k8s:openapi-gen=true,x-kubernetes-get-columns="FIELD_TITLE:field.path",x-kubernetes-describe-columns="FIELD_TITLE:field.path,OTHER_FIELD_TITLE:other.field.path"
type Type struct {
  // +k8s:openapi-gen=x-kubernetes-patch-strategy="merge",x-kubernetes-patch-merge-key="name"
  Field Type `patchStrategy:"merge" patchMergeKey:"name"`
}
```

The long term plan is to remove the `patchStrategy` and `patchMergeKey` struct tags once all code has been updated to use the swagger spec instead.  Until then, comments should be added for the struct tags `patchStrategy` and `patchStrategy`.  Some verification process will need to be written to make sure the struct tags and comments remain in sync.

### PR DESCRIPTION
**Pull request by mbohlool**
fixes #40818

### CODE REVIEW

**Comment by pwittrock** [comment]
*[cmd/libs/go2idl/openapi-gen/generators/openapi_test.go, line 113 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959#-Kc5swHX_veKjgFGMJyx:-Kc5swHX_veKjgFGMJyy:bkgfh8g) ([raw file](https://github.com/kubernetes/kubernetes/blob/65a2641dc40f09062fbec28766119d7d263c877d/cmd/libs/go2idl/openapi-gen/generators/openapi_test.go#L113)):*
> ```Go
> 	// a member with an extension
> 	// +k8s:openapi-gen=x-kubernetes-member-tag:member_test
> 	WithExtention string
> ```

nit: WithExtension?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959)*

**Comment by pwittrock** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959)*

**Comment by pwittrock** [comment]
Reviewed 2 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959)*

**Comment by pwittrock** [comment]
/lgtm

Feel free to reapply after fixing comment.

**Comment by mbohlool** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[cmd/libs/go2idl/openapi-gen/generators/openapi_test.go, line 113 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959#-Kc5swHX_veKjgFGMJyx:-Kc60Zldc_0F62HE42b_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/65a2641dc40f09062fbec28766119d7d263c877d/cmd/libs/go2idl/openapi-gen/generators/openapi_test.go#L113)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary><blockquote>

nit: WithExtension?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40959)*

**Comment by pwittrock** [comment]
@k8s-bot bazel test this issue #40295

File: cmd/libs/go2idl/openapi-gen/generators/openapi.go [hunk 1]
```diff
@@ -40,15 +40,21 @@ const tagOptional = "optional"
...
+	tagValueFalse      = "false"
+	tagExtensionPrefix = "x-kubernetes-"
 )
 
-func hasOpenAPITagValue(comments []string, value string) bool {
+func getOpenAPITagValue(comments []string) []string {
 	tagValues := types.ExtractCommentTags("+", comments)[tagName]
 	if tagValues == nil {
-		return false
+		return []string{}
```
Comment (lavalamp) [inline cmd/libs/go2idl/openapi-gen/generators/openapi.go]: can this entire function be `return types.ExtractCommentTags("+", comments)[tagName]`?

**Review by lavalamp — COMMENTED** [review]
very minor nits. :)

File: cmd/libs/go2idl/openapi-gen/generators/openapi.go (hunk 1 above)
Comment (mbohlool) [inline cmd/libs/go2idl/openapi-gen/generators/openapi.go]: You are right. Done.

File: cmd/libs/go2idl/openapi-gen/README [hunk 2]
```diff
@@ -2,3 +2,9 @@
 
 - To generate definition for a specific type or package add "+k8s:openapi-gen=true" tag to the type/package comment lines.
 - To exclude a type or a member from a tagged package/type, add "+k8s:openapi-gen=false" tag to the comment lines.
+
+# OpenAPI Extensions
+OpenAPI spec can have extension on types. To define one or more extension on a type or its member
+add "+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE" to the comment lines before type/member. A type/member can
+have multiple extensions. The rest of the line in the comment will be used as $VALUE so there is no need to
+escape or quote the value string.
```
Comment (mbohlool) [inline cmd/libs/go2idl/openapi-gen/README]: Done.

File: cmd/libs/go2idl/openapi-gen/README [hunk 3]
```diff
@@ -2,3 +2,9 @@
 
 - To generate definition for a specific type or package add "+k8s:openapi-gen=true" tag to the type/package comment lines.
 - To exclude a type or a member from a tagged package/type, add "+k8s:openapi-gen=false" tag to the comment lines.
+
+# OpenAPI Extensions
+OpenAPI spec can have extension on types. To define one or more extension on a type or its member
```
Comment (mbohlool) [inline cmd/libs/go2idl/openapi-gen/README]: Done.

**Comment by mbohlool** [comment]
@lavalamp PTAL.

**Comment by lavalamp** [comment]
/lgtm
/approve

sorry, thought I approved this a couple days ago!

**Pull request merged**